ethereum/eth/src/jmh/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldStateDownloaderBenchmark.java

```diff
@@ -0,0 +1,145 @@
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+@State(Scope.Thread)
+public class WorldStateDownloaderBenchmark {
+
+  private final BlockDataGenerator dataGen = new BlockDataGenerator();
+  private Path tempDir;
+  private BlockHeader blockHeader;
+  private final MetricsSystem metricsSystem = new NoOpMetricsSystem();
+  private WorldStateDownloader worldStateDownloader;
+  private WorldStateStorage worldStateStorage;
+  private RespondingEthPeer peer;
+  private Responder responder;
+  private TaskQueue<NodeDataRequest> pendingRequests;
+  private StorageProvider storageProvider;
+  private EthProtocolManager ethProtocolManager;
+  private InMemoryKeyValueStorage remoteKeyValueStorage;
+
+  @Setup(Level.Invocation)
+  public void setUpUnchangedState() throws Exception {
+    final SynchronizerConfiguration syncConfig =
+        new Builder().worldStateHashCountPerRequest(200).build();
+    final Hash stateRoot = createExistingWorldState();
+    blockHeader = new BlockHeaderTestFixture().stateRoot(stateRoot).buildHeader();
+
+    tempDir = Files.createTempDir().toPath();
+    ethProtocolManager =
+        EthProtocolManagerTestUtil.create(
+            new EthScheduler(
+                syncConfig.downloaderParallelism(),
+                syncConfig.transactionsParallelism(),
+                syncConfig.computationParallelism(),
+                metricsSystem));
+
+    peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager, blockHeader.getNumber());
+
+    final EthContext ethContext = ethProtocolManager.ethContext();
+    storageProvider = RocksDbStorageProvider.create(tempDir.resolve("database"), metricsSystem);
+    worldStateStorage = storageProvider.createWorldStateStorage();
+
+    pendingRequests =
+        RocksDbTaskQueue.create(
+            tempDir.resolve("fastsync"),
+            NodeDataRequest::serialize,
+            NodeDataRequest::deserialize,
+            metricsSystem);
+    worldStateDownloader =
+        new WorldStateDownloader(
+            ethContext,
+            worldStateStorage,
+            pendingRequests,
+            syncConfig.getWorldStateHashCountPerRequest(),
+            syncConfig.getWorldStateRequestParallelism(),
+            syncConfig.getWorldStateRequestMaxRetries(),
+            metricsSystem);
+  }
+
+  private Hash createExistingWorldState() {
+    // Setup existing state
+    remoteKeyValueStorage = new InMemoryKeyValueStorage();
+    final WorldStateStorage storage = new KeyValueStorageWorldStateStorage(remoteKeyValueStorage);
+    final WorldStateArchive worldStateArchive = new WorldStateArchive(storage);
+    final MutableWorldState worldState = worldStateArchive.getMutable();
+
+    dataGen.createRandomAccounts(worldState, 10000);
```

Comment:
Is JMH the right place to try larger states?  It wold be interesting to see if the performance decline is linear or exponential, or worse, so we could do world states at 1,000, 10,000, and 100,000, and 1,000,000 to see what the curve is.

Reply:
JMH turned out to be a fairly poor fit for this type of test - we could parameterise it so we could generate larger states but the runtime is already pretty brutal because we wind up having to recreate that state for every benchmark run and JMH wants to do a lot of runs to check consistency.  This will give us some rough numbers to work off for now but to really drill into details a controlled network benchmark running the full node and a single 10-20 minute sync will give us something we can drill into that kind of detail on.

That said, in looking at Rinkeby fast syncs, the rate that we complete nodes is very consistent regardless of how many we've imported already. So it looks to be linear based on the number of nodes.